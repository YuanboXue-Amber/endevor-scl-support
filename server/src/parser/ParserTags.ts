import { isNullOrUndefined } from "util";
import { ITokenizedString } from "./Tokenizer";
import { DiagnosticSeverity } from "vscode-languageserver";
import { SCLDocument, SCLstatement } from "../documents/SCLDocument";
import { QUICKFIX_UPPERCASE_MSG } from "../codeActionProvider";

export const ParserTags = {
    ACTION: "ACTion",
    ADD: "ADD",
    ARCHIVE: "ARChive",
    BUILD: "BUIld",
    COPY: "COPy",
    DDNAME: "DDName",
    DELETE: "DELete",
    DSNAME: "DSName",
    ELEMENT: "ELEment",
    ENVIRONMENT: "ENVironment",
    FILE: "FILe",
    FROM: "FROm",
    GENERATE: "GENerate",
    LIST: "LISt",
    MEMBER: "MEMber",
    MOVE: "MOVe",
    NUMBER: "NUMber",
    OPTIONS: "OPTions",
    OPTION: "OPTion",
    PATH: "PATH",
    PRINT: "PRInt",
    RESTORE: "REStore",
    RETRIEVE: "RETrieve",
    SET: "SET",
    SIGNIN: "SIGnin",
    STAGE: "STAge",
    STOPRC: "STOprc",
    SUBSYSTEM: "SUBSystem",
    SYSTEM: "SYStem",
    THROUGH: "THRough",
    TRANSFER: "TRAnsfer",
    TO: "TO",
    TYPE: "TYPe",
    UPDATE: "UPDate",
    USSFILE: "USSFILE",
    VALIDATE: "VALidate",
    WHERE: "WHEre",
    CCID: "CCID",
    COMMENT: "COMment",
    NEW: "NEW",
    VERSION: "VERsion",
    IF: "if",
    PRESENT: "present",
    INPUT: "input",
    SOURCE: "source",
    OVERRIDE: "OVErride",
    SIGNOUT: "SIGNOut",
    BYPASS: "BYPass",
    PROCESSOR: "PROcessor",
    GROUP: "GROup",
    EQUAL: "EQual",
    EQUALSIGN: "=",
    AUTOGEN: "AUTogen",
    SPAN: "SPAN",
    NONE: "NONe",
    ALL: "ALL",
    SYSTEMS: "SYStems",
    SUBSYSTEMS: "Subsystems",
};

/**
 * Get field value based on field name
 *
 * @param {string} tag field name in ParserTag classs, eg. ADD
 * @returns {(string | undefined)}
 */
function getString(tag: string): string | undefined {
    const descriptor = Object.getOwnPropertyDescriptor(ParserTags, tag);
    if (isNullOrUndefined(descriptor)) {
        return undefined;
    }
    return descriptor.value;
}


/**
 *
 * Match an input scl piece to a keyword specified by parserTag.
 *
 * @export
 * @param {ITokenizedString} inputSCLtoken
 * @param {string} parserTag
 * @returns {boolean} true when match false not
 */
export function matchWithoutDiagnose(
    inputSCLtoken: ITokenizedString,
    parserTag: string): boolean {

    const inputSCLpart = inputSCLtoken.value;
    let keyword = getString(parserTag);
    if (isNullOrUndefined(keyword)) {
        return false;
    }

    // might be operator
    if (keyword === inputSCLtoken.value) {
        return true;
    }

    const kwMatchArray = keyword.match(/([A-Z]*)([a-zA-Z]*)/);
    if (isNullOrUndefined(kwMatchArray)) {
        return false;
    }
    const mandatoryText = kwMatchArray[1];
    const optionalText = kwMatchArray[2].toUpperCase();

    const sclkey = inputSCLpart.trim().toUpperCase();
    if (!sclkey.startsWith(mandatoryText)) {
        return false;
    }

    // compose a regex from parsetag. Eg. for APPROVER it is APP(ROVER|ROVE|ROV|RO|R|\b)
    const composeRegex = ((mandatoryText: string, optionalText: string) => {
        let regexString = mandatoryText + "(";
        for (let i = optionalText.length; i > 0; -- i) {
            regexString += optionalText.substr(0, i) + "|";
        }
        regexString += "\\b)";
        return regexString;
    });
    const regex = new RegExp(composeRegex(mandatoryText, optionalText));
    const textMatchArray = sclkey.match(regex);
    if (!isNullOrUndefined(textMatchArray) && textMatchArray[0] === sclkey) {
        return true;
    }
    return false;
}

/**
 * Match an input scl piece to a keyword specified by parserTag.
 * When matched, perform a syntax diagnose on if the keyword is uppercased or not
 *
 * @export
 * @param {ITokenizedString} inputSCLtoken the token to be matched, which is in statement
 * @param {string} parserTag field name in ParserTag classs, eg. ADD
 * @param {SCLstatement} statement Only changed when return true: the statement where keyword uppercase warning diagnose will be pushed to
 * @param {SCLDocument} document Only changed when return true: the document where statement is in
 * @returns {boolean} true only when the field match the keyword represent by parserTag
 */
export function match(
    inputSCLtoken: ITokenizedString,
    parserTag: string,
    statement: SCLstatement,
    document: SCLDocument ): boolean {

    const ismatch = matchWithoutDiagnose(inputSCLtoken, parserTag);
    if (ismatch) {
        keywordUppercaseDiagnose(inputSCLtoken, statement, document);
    }
    return ismatch;
}

/**
 * Push a warning to a keyword that is not uppercased
 *
 * @param {ITokenizedString} keywordInSource
 * @param {SCLstatement} statement
 * @param {SCLDocument} document
 */
function keywordUppercaseDiagnose(
    keywordInSource: ITokenizedString,
    statement: SCLstatement,
    document: SCLDocument) {

    if (keywordInSource.value.toUpperCase() !== keywordInSource.value) {
        document.pushDiagnostic(
            keywordInSource,
            statement,
            DiagnosticSeverity.Warning,
            QUICKFIX_UPPERCASE_MSG + '\nLowercased keyword might cause the scl action to fail when submitted',
            );
    }
}