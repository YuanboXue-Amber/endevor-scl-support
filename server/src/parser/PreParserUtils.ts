import path = require('path');
import fs = require('fs');
import { isNullOrUndefined } from 'util';
import { ITokenizedString } from './Tokenizer';

export interface Inode {
    keyword?: string; // for keywords, if multiple, separated by comma
    maxLen?: number; // for value
    required?: boolean;
    isFROM?: boolean;
    isTO?: boolean;
    next?: Inode[];
    nogoback?: boolean; // when traverse on nodes, don't check this node
    specialValue?: boolean; // values in (,). Note all maxLen 255 and 768 are automaticaly special values,
    dateTimeValue?: boolean; // value with date time format
    parent?: Inode;

    requireNext?: boolean; // require child to be specified
}

export let SETtree: Inode;
export let ADDtree: Inode;
export let GENERATEtree: Inode;
export let DEFINEPACKAGEtree: Inode;

export let UPDATEtree: Inode;
export let DELETEtree: Inode;
export let MOVEtree: Inode;
export let RETRIEVEtree: Inode;
export let SIGNINtree: Inode;
export let TRANSFERtree: Inode;
export let APPROVEtree: Inode;
export let DENYtree: Inode;
export let BACKINtree: Inode;
export let BACKOUTtree: Inode;
export let CASTtree: Inode;
export let EXECUTEtree: Inode;
export let RESETtree: Inode;
export let COMMITtree: Inode;
export let LISTtree: Inode;

export function prepareTrees() {
    SETtree = composeTreeFromJSON(prepareJSON("SETtree.json"));
    ADDtree = composeTreeFromJSON(prepareJSON("ADDtree.json"));
    GENERATEtree = composeTreeFromJSON(prepareJSON("GENERATEtree.json"));
    DEFINEPACKAGEtree = composeTreeFromJSON(prepareJSON("DEFINEPACKAGEtree.json"));
    APPROVEtree = composeTreeFromJSON(prepareJSON("APPROVEtree.json"));
    DENYtree = composeTreeFromJSON(prepareJSON("DENYtree.json"));
    BACKINtree = composeTreeFromJSON(prepareJSON("BACKINtree.json"));
    BACKOUTtree = composeTreeFromJSON(prepareJSON("BACKOUTtree.json"));
    CASTtree = composeTreeFromJSON(prepareJSON("CASTtree.json"));
    DEFINEPACKAGEtree = composeTreeFromJSON(prepareJSON("DEFINEPACKAGEtree.json"));
    EXECUTEtree = composeTreeFromJSON(prepareJSON("EXECUTEtree.json"));
    RESETtree = composeTreeFromJSON(prepareJSON("RESETtree.json"));
    COMMITtree = composeTreeFromJSON(prepareJSON("COMMITtree.json"));
    UPDATEtree = composeTreeFromJSON(prepareJSON("UPDATEtree.json"));
    DELETEtree = composeTreeFromJSON(prepareJSON("DELETEtree.json"));
    MOVEtree = composeTreeFromJSON(prepareJSON("MOVEtree.json"));
    RETRIEVEtree = composeTreeFromJSON(prepareJSON("RETRIEVEtree.json"));
    SIGNINtree = composeTreeFromJSON(prepareJSON("SIGNINtree.json"));
    TRANSFERtree = composeTreeFromJSON(prepareJSON("TRANSFERtree.json"));
    LISTtree = composeTreeFromJSON(prepareJSON("LISTtree.json"));
}

function prepareJSON(jsonFileName: string): Inode {
    const jsonPath = path.resolve(__dirname, "../parser/syntaxTrees/"+ jsonFileName);
    const jsonStr = fs.readFileSync(jsonPath, "UTF-8");
    return JSON.parse(jsonStr);
}

function composeTreeFromJSON(jsonRootNode: Inode): Inode {
    const fillInParent = (jsonParentNode: Inode) => {
        if (isNullOrUndefined(jsonParentNode.next))
            return;

        jsonParentNode.next.forEach((childNode) => {
            childNode.parent = jsonParentNode;
            fillInParent(childNode);
        });
    };

    fillInParent(jsonRootNode);
    return jsonRootNode;
}

export function dispatcher(scl: ITokenizedString[]): Inode | undefined {
    if (scl.length >= 1) {
        switch (true) {
            case match(scl[0].value, "ADD"):
                return ADDtree;
            case match(scl[0].value, "GENerate"):
                return GENERATEtree;
            case match(scl[0].value, "DEFine"):
                return DEFINEPACKAGEtree;
            case match(scl[0].value, "SET"):
                return SETtree;
            case match(scl[0].value, "APPRove"):
                return APPROVEtree;
            case match(scl[0].value, "BACKIn"):
                return BACKINtree;
            case match(scl[0].value, "BACKOut"):
                return BACKOUTtree;
            case match(scl[0].value, "CASt"):
                return CASTtree;
            case match(scl[0].value, "COMMit"):
                return COMMITtree;
            case match(scl[0].value, "DENY"):
                return DENYtree;
            case match(scl[0].value, "EXECUTE"):
                return EXECUTEtree;
            case match(scl[0].value, "RESet"):
                return RESETtree;
            case match(scl[0].value, "DELete"):
                return DELETEtree;
            case match(scl[0].value, "LISt"):
                return LISTtree;
            case match(scl[0].value, "UPDate"):
                return UPDATEtree;
            case match(scl[0].value, "MOVe"):
                return MOVEtree;
            case match(scl[0].value, "RETrieve"):
                return RETRIEVEtree;
            case match(scl[0].value, "SIGnin"):
                return SIGNINtree;
            case match(scl[0].value, "TRAnsfer"):
                return TRANSFERtree;
            default:
                break;
        }
    }
    return undefined;
}

export function match(wordToMatch: string, pattern: string) {
    if (wordToMatch.toUpperCase() === pattern.toUpperCase()) {
        return true;
    }

    const kwMatchArray = pattern.match(/([A-Z]*)([a-zA-Z]*)/);
    if (isNullOrUndefined(kwMatchArray)) {
        return false;
    }
    const mandatoryText = kwMatchArray[1];
    const optionalText = kwMatchArray[2].toUpperCase();

    const sclkey = wordToMatch.trim().toUpperCase();
    if (!sclkey.startsWith(mandatoryText)) {
        return false;
    }

    // compose a regex from parsetag. Eg. for APPROVER it is APP(ROVER|ROVE|ROV|RO|R|\b)
    const composeRegex = ((mandatoryText: string, optionalText: string) => {
        let regexString = mandatoryText + "(";
        for (let i = optionalText.length; i > 0; -- i) {
            regexString += optionalText.substr(0, i) + "|";
        }
        regexString += "\\b)";
        return regexString;
    });
    const regex = new RegExp(composeRegex(mandatoryText, optionalText));
    const textMatchArray = sclkey.match(regex);
    if (!isNullOrUndefined(textMatchArray) && textMatchArray[0] === sclkey) {
        return true;
    }
    return false;
}
