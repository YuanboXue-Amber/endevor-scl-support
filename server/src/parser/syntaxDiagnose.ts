import { TextDocument, Diagnostic, DiagnosticSeverity } from "vscode-languageserver";
import { TokenizedString, Tokenizer } from "./tokenizer";
import { match } from "./parserTags";
import { isNullOrUndefined } from "util";
import { parseSetAction } from "./syntaxTrees/setCommands";

interface ExampleSettings {
    maxNumberOfProblems: number;
}
export const defaultSettings: ExampleSettings = { maxNumberOfProblems: 1000 };

export class SyntaxDiagnose {
    textDocument: TextDocument;
    hasDiagnosticRelatedInformationCapability: boolean;
    textDocContent: string;
    tokenizingTextContent: Tokenizer;
    diagnostics: Diagnostic[] = [];

    constructor(textDocument: TextDocument, hasDiagnosticRelatedInformationCapability: boolean) {
        this.textDocument = textDocument;
        this.hasDiagnosticRelatedInformationCapability = hasDiagnosticRelatedInformationCapability;
        this.textDocContent = textDocument.getText();
        this.tokenizingTextContent = new Tokenizer(this.textDocContent);
        this.diagnostics = [];
        this.SplitAndDispatchSCL();
    }

    /**
     * For the entire file content, read one by one until we met the end of an SCL, send it to process.
     * Then continue to read the next one.
     *
     * @private
     * @memberof SyntaxDiagnose
     */
    private SplitAndDispatchSCL() {
        let token: TokenizedString = this.tokenizingTextContent.peekNext();
        let thisSCL: TokenizedString[] = [];
        while(!isNullOrUndefined(token) && !token.is_eof) {
            token = this.tokenizingTextContent.readNext();
            if (token.is_eof) {
                if (thisSCL.length > 0) {
                    if (match(thisSCL[0], "T_SET", this)) {
                        this.processSETSCL(thisSCL);
                        break;
                    }
                    // no need to wait for scls that are not a SET xxx
                    process.nextTick(this.dispatchSCL, thisSCL);
                }
                break;
            }
            thisSCL.push(token);

            if (token.is_eoStatement) {
                const sclToBeProcess = thisSCL;
                thisSCL = [];
                if (match(sclToBeProcess[0], "T_SET", this)) {
                    this.processSETSCL(sclToBeProcess);
                    continue;
                }
                // no need to wait for scls that are not a SET xxx
                process.nextTick(this.dispatchSCL, sclToBeProcess);
            }
        }
    }

    private processSETSCL(currSCL: TokenizedString[]) {
        // currSCL[0] is SET for sure

        if (currSCL.length < 2) {
            this.pushDiagnostic(DiagnosticSeverity.Error, currSCL[0],
                'SET should be followed by the following keywords: ACTION, BUILD, FROM, OPTIONS, STOPRC, TO, WHERE');
            return;
        }

        let i = 1;
        let token = currSCL[i];
        switch (true) {
            case match(token, "T_ACTION", this):
                parseSetAction(this, currSCL, i);
                return;
            case match(token, "T_BUILD", this):

                break;
            case match(token, "T_FROM", this):

                break;
            case match(token, "T_OPTIONS", this):

                break;
            case match(token, "T_STOPRC", this):

                break;
            case match(token, "T_TO", this):

                break;
            case match(token, "T_WHERE", this):

                break;
            default:
                this.pushDiagnostic(DiagnosticSeverity.Error, token,
                    `Invalid keyword \"${token.value}\" specified after SET`,
                    'SET should be followed by the following keywords: ACTION, BUILD, FROM, OPTIONS, STOPRC, TO, WHERE');
                return;
        }
    }

    /**
     * Send the scl to its own syntax validate tree based on the first keyword.
     *
     * @param {TokenizedString[]} currSCL
     * @memberof SyntaxDiagnose
     */
    dispatchSCL = ((currSCL: TokenizedString[]) => {

    });

    /**
     * Get the diagnostic info and push it into this.diagnostics
     *
     * @param {DiagnosticSeverity} severity
     * @param {TokenizedString} token
     * @param {string} message
     * @param {string} [relatedMsg]
     * @memberof SyntaxDiagnose
     */
    public pushDiagnostic(severity: DiagnosticSeverity, token: TokenizedString, message: string, relatedMsg?: string) {
        if (this.diagnostics.length > defaultSettings.maxNumberOfProblems) {
            return;
        }

        let diagnostic: Diagnostic = {
            severity: severity,
            range: {
                start: this.textDocument.positionAt(token.startPos),
                end: this.textDocument.positionAt(token.startPos + token.value.length)
            },
            message: message,
            source: 'Endevor SCL extension'
        };
        if (this.hasDiagnosticRelatedInformationCapability && !isNullOrUndefined(relatedMsg)) {
            diagnostic.relatedInformation = [
                {
                    location: {
                        uri: this.textDocument.uri,
                        range: Object.assign({}, diagnostic.range)
                    },
                    message: relatedMsg,
                }
            ];
        }
        this.diagnostics.push(diagnostic);
    }
}

