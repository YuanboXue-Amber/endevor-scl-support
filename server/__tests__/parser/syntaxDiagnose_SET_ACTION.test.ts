import { SyntaxDiagnose } from '../../src/parser/syntaxDiagnose';
import { Position } from 'vscode-languageserver';

describe("Test a simple valid SET ACTION", () => {
    it("Should have no error", async () => {
        let sourceTextDocument: any = {
            getText() {
                return " SET ACTION VALIDATE . ";
            },
            uri: "mocked/uri"
        };

        const syntaxDiagnose: SyntaxDiagnose = new SyntaxDiagnose(sourceTextDocument, true);
        expect(syntaxDiagnose.diagnostics.length).toBe(0);
    });
});

describe("Test 3 simple valid SET ACTION", () => {
    it("Should have no error", async () => {
        let sourceTextDocument: any = {
            getText() {
                return " SET ACTION VALIDATE . \n      SET ACTION ADD . \nSET ACTION TRA . ";
            },
            uri: "mocked/uri"
        };

        const syntaxDiagnose: SyntaxDiagnose = new SyntaxDiagnose(sourceTextDocument, true);
        expect(syntaxDiagnose.diagnostics.length).toBe(0);
    });
});

describe("Test 4 simple INvalid SET ACTION", () => {
    it("Should have no error", async () => {
        let sourceTextDocument: any = {
            getText() {
                return " SET ACTION INVALID . SET \n  . SET ACTION RES. SET ACTION ADD";
            },
            positionAt(offset: number): Position {
                expect(offset).toMatchSnapshot();
                const mockPosition: Position = {
                    line: 1,
                    character: 2
                };
                return mockPosition;
            },
            uri: "mocked/uri"
        };

        const syntaxDiagnose: SyntaxDiagnose = new SyntaxDiagnose(sourceTextDocument, true);
        expect(syntaxDiagnose.diagnostics).toMatchSnapshot();
    });
});

describe("Test 2 simple INvalid SET ACTION, with lowercased keyword", () => {
    it("Should have no error", async () => {
        let sourceTextDocument: any = {
            getText() {
                return " SET ACTION res . SET ACTiON invalid";
            },
            positionAt(offset: number): Position {
                expect(offset).toMatchSnapshot();
                const mockPosition: Position = {
                    line: 1,
                    character: 2
                };
                return mockPosition;
            },
            uri: "mocked/uri"
        };

        const syntaxDiagnose: SyntaxDiagnose = new SyntaxDiagnose(sourceTextDocument, true);
        expect(syntaxDiagnose.diagnostics).toMatchSnapshot();
    });
});