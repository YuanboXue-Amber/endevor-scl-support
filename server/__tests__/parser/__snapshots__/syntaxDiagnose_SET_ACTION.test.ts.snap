// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test 2 simple INvalid SET ACTION, with lowercased keyword Should have no error 1`] = `12`;

exports[`Test 2 simple INvalid SET ACTION, with lowercased keyword Should have no error 2`] = `15`;

exports[`Test 2 simple INvalid SET ACTION, with lowercased keyword Should have no error 3`] = `22`;

exports[`Test 2 simple INvalid SET ACTION, with lowercased keyword Should have no error 4`] = `28`;

exports[`Test 2 simple INvalid SET ACTION, with lowercased keyword Should have no error 5`] = `29`;

exports[`Test 2 simple INvalid SET ACTION, with lowercased keyword Should have no error 6`] = `36`;

exports[`Test 2 simple INvalid SET ACTION, with lowercased keyword Should have no error 7`] = `
Array [
  Object {
    "message": "Keyword should be uppercased",
    "range": Object {
      "end": Object {
        "character": 2,
        "line": 1,
      },
      "start": Object {
        "character": 2,
        "line": 1,
      },
    },
    "relatedInformation": Array [
      Object {
        "location": Object {
          "range": Object {
            "end": Object {
              "character": 2,
              "line": 1,
            },
            "start": Object {
              "character": 2,
              "line": 1,
            },
          },
          "uri": "mocked/uri",
        },
        "message": "Lowercased keyword might cause the scl action to fail when submitted",
      },
    ],
    "severity": 2,
    "source": "Endevor SCL extension",
  },
  Object {
    "message": "Keyword should be uppercased",
    "range": Object {
      "end": Object {
        "character": 2,
        "line": 1,
      },
      "start": Object {
        "character": 2,
        "line": 1,
      },
    },
    "relatedInformation": Array [
      Object {
        "location": Object {
          "range": Object {
            "end": Object {
              "character": 2,
              "line": 1,
            },
            "start": Object {
              "character": 2,
              "line": 1,
            },
          },
          "uri": "mocked/uri",
        },
        "message": "Lowercased keyword might cause the scl action to fail when submitted",
      },
    ],
    "severity": 2,
    "source": "Endevor SCL extension",
  },
  Object {
    "message": "Invalid action \\"invalid\\" specified after SET ACTION",
    "range": Object {
      "end": Object {
        "character": 2,
        "line": 1,
      },
      "start": Object {
        "character": 2,
        "line": 1,
      },
    },
    "relatedInformation": Array [
      Object {
        "location": Object {
          "range": Object {
            "end": Object {
              "character": 2,
              "line": 1,
            },
            "start": Object {
              "character": 2,
              "line": 1,
            },
          },
          "uri": "mocked/uri",
        },
        "message": "Valid actions are: ADD, ARCHIVE, COPY, DELETE, GENERATE, LIST, MOVE, PRINT, RESTORE, RETRIEVE, SIGNIN, TRANSFER, UPDATE, VALIDATE",
      },
    ],
    "severity": 1,
    "source": "Endevor SCL extension",
  },
]
`;

exports[`Test 4 simple INvalid SET ACTION Should have no error 1`] = `12`;

exports[`Test 4 simple INvalid SET ACTION Should have no error 2`] = `19`;

exports[`Test 4 simple INvalid SET ACTION Should have no error 3`] = `29`;

exports[`Test 4 simple INvalid SET ACTION Should have no error 4`] = `30`;

exports[`Test 4 simple INvalid SET ACTION Should have no error 5`] = `45`;

exports[`Test 4 simple INvalid SET ACTION Should have no error 6`] = `46`;

exports[`Test 4 simple INvalid SET ACTION Should have no error 7`] = `58`;

exports[`Test 4 simple INvalid SET ACTION Should have no error 8`] = `61`;

exports[`Test 4 simple INvalid SET ACTION Should have no error 9`] = `
Array [
  Object {
    "message": "Invalid action \\"INVALID\\" specified after SET ACTION",
    "range": Object {
      "end": Object {
        "character": 2,
        "line": 1,
      },
      "start": Object {
        "character": 2,
        "line": 1,
      },
    },
    "relatedInformation": Array [
      Object {
        "location": Object {
          "range": Object {
            "end": Object {
              "character": 2,
              "line": 1,
            },
            "start": Object {
              "character": 2,
              "line": 1,
            },
          },
          "uri": "mocked/uri",
        },
        "message": "Valid actions are: ADD, ARCHIVE, COPY, DELETE, GENERATE, LIST, MOVE, PRINT, RESTORE, RETRIEVE, SIGNIN, TRANSFER, UPDATE, VALIDATE",
      },
    ],
    "severity": 1,
    "source": "Endevor SCL extension",
  },
  Object {
    "message": "Invalid keyword \\".\\" specified after SET",
    "range": Object {
      "end": Object {
        "character": 2,
        "line": 1,
      },
      "start": Object {
        "character": 2,
        "line": 1,
      },
    },
    "relatedInformation": Array [
      Object {
        "location": Object {
          "range": Object {
            "end": Object {
              "character": 2,
              "line": 1,
            },
            "start": Object {
              "character": 2,
              "line": 1,
            },
          },
          "uri": "mocked/uri",
        },
        "message": "SET should be followed by the following keywords: ACTION, BUILD, FROM, OPTIONS, STOPRC, TO, WHERE",
      },
    ],
    "severity": 1,
    "source": "Endevor SCL extension",
  },
  Object {
    "message": "Invalid scl. Expecting a space before end of statement operator \\".\\"",
    "range": Object {
      "end": Object {
        "character": 2,
        "line": 1,
      },
      "start": Object {
        "character": 2,
        "line": 1,
      },
    },
    "severity": 1,
    "source": "Endevor SCL extension",
  },
  Object {
    "message": "No end of statement operator \\".\\" specified",
    "range": Object {
      "end": Object {
        "character": 2,
        "line": 1,
      },
      "start": Object {
        "character": 2,
        "line": 1,
      },
    },
    "severity": 1,
    "source": "Endevor SCL extension",
  },
]
`;
